Add service module to write PolKit agents

Cu3PO42 opened the pull request:
This implements a new service module that enables users to write PolKit agents by integrating with `libpolkit` and `libpolkitagent`. A test configuration is added to the source.

**Status**: I've done everything I have planned.

<details>
<summary>Original description</summary>

This implements a new service module that enables users to write PolKit agents. To this end, I'm integrating with the existing `polkit-qt-1` library that is also used by the KDE agent.

A relatively minimal example configuration using this new service can be found [here](https://gist.github.com/Cu3PO42/12675e61f9e497d22313e12c09117852).

There's a minor TODO left around cleanup of the `Identity` objects that I haven't been able to get quite right due to my inexperience with Qt. I'm hoping you might have a clean solution for this.

</details>

Closes #271.

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,428 @@
+
+PolkitAgent::~PolkitAgent() {
+	// First, destroy all but the first request in the queue, so that the cancel
+	// method doesn't start a new one.
```
I would think its not unreasonable to handle a request right after cancelling a different one, especially if they're for different programs.

Also, when reloading the graph, preserving the old polkit session may be the most reasonable case, especially if the user wants to do something like edit their auth dialog with a real app running to test with.

Cu3PO42 replied:
Handling one request after another is the default behavior in `finishAuthenticationRequest`. The `cancelAuthenticationRequest` method calls `finishAuthenticationRequest` and therefore would start another request immediately.

However, we are in the destructor of the agent here, so there will be no-one to handle the queued requests after the destructor has run. Therefore I want to reject all currently queued requests and not start processing them. To this end, I start by rejecting all requests after the current one, so that when the cancel method runs, it won't have any more to start.

outfoxxed replied:
As for reloading see my comment below about using an internal singleton. you can wait for the destruction of the last engine generation with a ref to the backing agent to cancel all requests and deregister the agent, or some similar pattern.

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,428 @@
+			                                                : QString::number(uid);
+			QString icon;
+			if (pw && pw->pw_dir && *pw->pw_dir) {
+				icon = QString::fromUtf8(pw->pw_dir) + QDir::separator() + ".face.icon";
```
Is this a standard? I've heard of multiple paths to a user picture

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,428 @@
+		}
+	}
+
+	mSelectedIdentity = mIdentities.isEmpty() ? nullptr : mIdentities.first();
```
I'm not sure how much significance it has with polkit, but would it make sense for a user to pick any other identity if multiple are available?

outfoxxed replied:
Then we should probably expose that as an option

Cu3PO42 replied:
The user can set `selectedIdentity` at any time, even after authentication is already started. This is just a (in my opinion) reasonable default. The alternative would be to not set this and always make the user choose.

This behavior also mirrors what all existing PolKit agents that I'm aware of do.

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,428 @@
+
+		// In case of config reloads new objects are constructed before the old ones
+		// are destroyed. Therefore the new agent cannot take over the path still in
+		// use by the old agent.
```
Theres a couple differences in the approaches you could take:
- Reloadable uses heuristics to try to match an object in the old generation to one in the new generation. This is the system that allows allocated windows to be reused. You could use this if you wanted to only persist the polkit agent if the user placed it in the same spot in the same file across generations.
- PostReloadHook fires unconditionally after Reloadable and has no concept of the scene graph. Combining this with `findObjectGeneration` on the user facing wrapper objects from the current and previous generations will tell you if we're switching generations or the user created a second invalid agent. Avoid `currentGeneration`.

Either of these interfaces you would implement on the wrapper object and not the internal singleton.

I recommend the PostReloadHook approach but would accept either.

outfoxxed commented on src/services/polkit/flow.hpp:
```diff
@@ -0,0 +1,162 @@
+	/// The identity that will be used to authenticate.
+	/// 
+	/// Setting this will abort any ongoing authentication conversations and start a new one.
+	Q_PROPERTY(Identity* selectedIdentity READ selectedIdentity WRITE setSelectedIdentity NOTIFY selectedIdentityChanged);
```
"Setting" -> "Changing"

Reactive changes can cause a lot of "Setting" so this wording could be confusing

outfoxxed commented on src/services/polkit/flow.hpp:
```diff
@@ -0,0 +1,162 @@
+	/// An additional message to present to the user.
+	///
+	/// This may be used to show errors or supplementary information.
+	Q_PROPERTY(QString supplementaryMessage READ supplementaryMessage NOTIFY supplementaryChanged);
```
Add something along the lines of "See @@supplementaryIsError to determine if this should be displayed as an error message"

outfoxxed commented on src/services/polkit/flow.cpp:
```diff
@@ -0,0 +1,159 @@
+
+void AuthFlow::setSelectedIdentity(Identity* identity) {
+	if (mSelectedIdentity == identity) return;
+	if (!identity) return; // ignore null changes
```
I would print an error here with qmlWarning(this)

outfoxxed commented on src/services/polkit/identity.hpp:
```diff
@@ -0,0 +1,59 @@
+
+	// clang-format off
+	/// The Id of the identity. If the identity is a user, this is the user's uid. See @@isGroup.
+	Q_PROPERTY(id_t id READ id CONSTANT);
```
Please cast/rename to a qt primitive type in the interface. This will show up as `unknown` in the doc generator.

outfoxxed commented on src/services/polkit/listener.cpp:
```diff
@@ -0,0 +1,202 @@
+
+		identitySet.insert(polkit_identity_hash(identity));
+		identityVector.push_back(identity);
+		g_object_ref(identity);
```
I see we add a ref to the identities we're using. Add a comment if polkit takes care of derefing all the ones we dont' or do it if we need to

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,69 @@
+
+void PolkitAgent::classBegin() {
+	// Nothing to do here.
+}
```
just leave classbegin () {} in the header

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,69 @@
+namespace qs::service::polkit {
+PolkitAgent::PolkitAgent(QObject* parent): PostReloadHook(parent) {}
+
+PolkitAgent::~PolkitAgent() = default;
```
this can also go in the header

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,69 @@
+
+	emit isRegisteredChanged();
+	emit isActiveChanged();
+	emit flowChanged();
```
Another good site for bindable properties

outfoxxed replied:
They are field backed but when used with a binding callback the field acts as more of a cache than anything else. It also compares old/new values to see if anything changed and emit signals or not.

There are cases where it improves readability, generally the more values that depend on eachother the greater the improvement.

Here's an example of it being used with a binding: https://github.com/quickshell-mirror/quickshell/blob/3e2ce40b18af943f9ba370ed73565e9f487663ef/src/wayland/wlr_layershell/wlr_layershell.cpp#L18

And heres an example of it being used as a field that just so happens to emit signals automatically on prop change: https://github.com/quickshell-mirror/quickshell/blob/3e2ce40b18af943f9ba370ed73565e9f487663ef/src/services/notifications/notification.cpp#L113

Up to you where you use it but it does generally make things clearer.

outfoxxed commented on src/services/polkit/agentimpl.hpp:
```diff
@@ -0,0 +1,49 @@
+	PolkitAgent* qmlAgent = nullptr;
+
+	AuthFlow* activeFlow = nullptr;
+	std::deque<AuthRequest*> queuedRequests;
```
It seems like we're accepting every request preemptively and then keeping them waiting. If we don't do that does polkit queue them or cancel them? Should we be leaving that to polkit?

Cu3PO42 commented on src/services/polkit/identity.cpp:
```diff
@@ -0,0 +1,62 @@
+
+Identity::~Identity() = default;
+
+Identity* Identity::fromPolkitIdentity(PolkitIdentity* identity) {
```
I thought I set a parent for them, but I didn't previously. I made them a child of the AuthFlow to take care of this.

Cu3PO42 commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,69 @@
+
+	if (mPath.isEmpty()) mPath = "/org/quickshell/Polkit";
+
+	PolkitAgentImpl::tryGetOrCreate(this);
```
This must be a case of not seeing the forest for the trees. This is indeed much simpler and I changed it. That said, the post reload hook always run before the destructor of the previous generation in my testing.

Cu3PO42 commented:
Thank you for the extensive review! I addressed most of your comments. I also introduced a RAII helper `GObjectRef` to make handling of GObject objetcs less error prone. I still need to look into bindable properties.

Cu3PO42 commented:
I rebased onto current master, squashed and fixed a small bug in the reloading logic. I'm currently trying to build a more full-featured example in a fork of Noctalia.

outfoxxed commented on src/services/polkit/qml.cpp:
```diff
@@ -0,0 +1,49 @@
+
+QBindable<AuthFlow*> PolkitAgent::flow() { return &this->bFlow; }
+QBindable<bool> PolkitAgent::isActive() { return &this->bIsActive; }
+QBindable<bool> PolkitAgent::isRegistered() { return &this->bIsRegistered; }
```
These can go in the header

outfoxxed commented on src/services/polkit/gobjectref.hpp:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	// We do handle self-assignment in a more general case by checking the
+	// included pointers rather than the wrapper objects themselves.
```
The comment says this which makes sense but it looks like we dont ever compare the gobjectref content pointers before unreffing?

outfoxxed replied:
Whoops yeah. My misread.

outfoxxed commented:
Generally my policy is one-line getter functions can go in the header. Takeover change looks fine here.

outfoxxed reviewed (changes requested):
Test issues discovered:
1. Polkit agent does not attempt to register itself if it failed and then the currently active agent leaves the bus. See notifications.
2. Debug log levels are still enabled
3. Supplementary error message is never populated for bad passwords

Also noticed: for bindables use Qt::beginPropertyUpdateGroup / end update group to not send useless updates.

Cu3PO42 commented:
> Test issues discovered:
> 
>     1. Polkit agent does not attempt to register itself if it failed and then the currently active agent leaves the bus. See notifications.

This is true. I didn't tackle that before since it seemed to add significant complexity for relatively little advantage. However, I see the `QDbusServiceWatcher` actually makes this quite simple. Will do.

>     2. Debug log levels are still enabled

My bad. I misunderstood how the logging categories worked. I thought the default log level was a runtime switch, not in the code.

>     3. Supplementary error message is never populated for bad passwords

True. That was an oversight when designing `clearState`. It never came up for me in testing since my PAM config starts with fingerprints and the `show-info` from that overwrites any error message on restart anyway. I'll keep the supplementary message in that case. (Side note: I really wish PAM had better structuring to conversations.)


> Also noticed: for bindables use Qt::beginPropertyUpdateGroup / end update group to not send useless updates.

Aha! I was looking for something like that, thanks!

outfoxxed commented:
For the log levels just add QtWarningMsg to the logging category declaration to set the default. QT_LOGGING_RULES is then usable as a runtime switch.

Cu3PO42 commented:
That should be everything. I have removed the entire `clearState` method in `Flow` intentionally. It previously reset the supplementary message and input request. We want to explicitly keep the supplementary message; and input request is already reset in `submit`, which always comes before `clearState` was called in `onComplete`.
